hail/python/hailtop/utils/os.py

```diff
@@ -0,0 +1,130 @@
+import os
+import shutil
+import glob
+from functools import wraps
+
+from .utils import blocking_to_async, flatten
+
+
+wildcards = ('*', '?', '[', ']', '{', '}')
+
+
+# need a custom escape because escaped characters are not treated properly with glob.escape
+# and fnmatch doesn't work with escaped characters like \?
+def escape(path):
+    new_path = []
+    n = len(path)
+    i = 0
+    while i < n:
+        if i < n - 1 and path[i] == '\\' and path[i + 1] in wildcards:
+            new_path.append('[')
+            new_path.append(path[i + 1])
+            new_path.append(']')
+            i += 2
+            continue
+
+        new_path.append(path[i])
+        i += 1
+    return ''.join(new_path)
+
+
+def contains_wildcard(c):
+    i = 0
+    n = len(c)
+    while i < n:
+        if i < n - 1 and c[i] == '\\' and c[i + 1] in wildcards:
+            i += 2
+            continue
+        elif c[i] in wildcards:
+            return True
+        i += 1
+    return False
+
+
+def unescape_escaped_wildcards(c):
+    new_c = []
+    i = 0
+    n = len(c)
+    while i < n:
+        if i < n - 1 and c[i] == '\\' and c[i + 1] in wildcards:
+            new_c.append(c[i + 1])
+            i += 2
+            continue
+        new_c.append(c[i])
+        i += 1
+    return ''.join(new_c)
+
+
+def prefix_wout_wildcard(c):
+    new_c = []
+    i = 0
+    n = len(c)
+    while i < n:
+        if i < n - 1 and c[i] == '\\' and c[i + 1] in wildcards:
+            new_c.append(c[i + 1])
+            i += 2
+            continue
+        elif c[i] in wildcards:
+            return ''.join(new_c)
+        new_c.append(c[i])
+        i += 1
+    return ''.join(new_c)
```

Comment:
Can we leave these four functions out until we need them for something like `glob`?